### PR TITLE
feat: add imagenet distributed training example

### PR DESCRIPTION
**Pull request by liamcli**
Add new distributed training example on ImageNet using a recently discovered NAS architecture from 
https://arxiv.org/abs/2004.07802.

# Test Plan
- UI change:
  - [ ] add screenshots
  - [ ] React? build & check storybooks
- [ ] user-facing api change: modify documentation and examples
- [ ] user-facing api change: add the "User-facing API Change" label
- [ ] bug fix: add regression test
- [ ] bug fix: determine if there are other similar bugs in the codebase
- [ ] new feature: add test coverage for any user-facing aspects
- [ ] refactor: maintain existing code coverage

### CODE REVIEW

File: examples/experimental/imagenet_gaea/gcp_bucket_helper.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+def list_blobs(storage_client, bucket_name, prefix=None):
+    """Lists all the blobs in the bucket."""
+    blobs = storage_client.list_blobs(bucket_name, prefix=prefix)
+    paths = []
+    for b in blobs:
+        paths.append(b.name)
+    return paths
+
+
+def download_blob(storage_client, bucket_name, source_blob_name, destination_file_name, n_retries=4, max_backoff=8):
```
Comment (aaron276h) [inline examples/experimental/imagenet_gaea/gcp_bucket_helper.py:15]: FYI we have https://github.com/determined-ai/determined/blob/e0dd3e51f33e441fe4b6a8aa7cb3591f204f2af1/harness/determined/util.py#L19 which looks fairly similar.

Comment (liamcli) [reply examples/experimental/imagenet_gaea/gcp_bucket_helper.py:15]: Thanks for the pointer!  I wound up implementing my own because I wanted to download to the filesystem directly.

File: examples/experimental/imagenet_gaea/gcp_bucket_helper.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        raise
+
+
+#if __name__=="__main__":
+    bucket_name = 'determined-ai-datasets'
+    prefix = 'imagenet/imagenet/validation'
+    storage_client = storage.Client()
+    img_paths = list_blobs(storage_client, bucket_name, prefix)
+    print(img_paths[0:10])
+    download_blob(storage_client, bucket_name, img_paths[2], os.path.join('/tmp',img_paths[2]))
```
Comment (katport) [inline examples/experimental/imagenet_gaea/gcp_bucket_helper.py:45]: I'm a bit confused about this chunk? Is this meant to be commented out?

File: examples/experimental/imagenet_gaea/model_def.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+        step_mode = LRScheduler.StepMode.MANUAL_STEP
+        return LRScheduler(self.scheduler, step_mode=step_mode)
+
+    def train_batch(
+        self, batch: Any, model: nn.Module, epoch_idx: int, batch_idx: int
+    ) -> Dict[str, torch.Tensor]:
+
+        if batch_idx==0 or self.last_epoch_idx < epoch_idx:
+            self.scheduler.step()
+            current_lr = self.scheduler.get_lr()[0]
```
Comment (katport) [inline examples/experimental/imagenet_gaea/model_def.py:172]: This LR scheduler should have a function called get_last_lr() which pulls the correct lr vs get_lr() which can also calculate the LRs.

File: examples/experimental/imagenet_gaea/gcp_bucket_helper.py (hunk 2 above)
Comment (liamcli) [reply examples/experimental/imagenet_gaea/gcp_bucket_helper.py]: This was used for my own testing purposes.  I now use the function aaron pointed to for downloading from GCP so this file has been removed.

File: examples/experimental/imagenet_gaea/model_def.py (hunk 3 above)
Comment (liamcli) [reply examples/experimental/imagenet_gaea/model_def.py]: Thanks for pointing that out.  Changed to get_last_lr.

File: examples/experimental/imagenet_nas_arch_pytorch/model_def.py [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+    'swish': Swish,
+    'hswish': HSwish
+}
+
+class ImageNetTrial(PyTorchTrial):
+    def __init__(self, context: det.TrialContext) -> None:
+        self.context = context
+        self.data_config = context.get_data_config()
+        self.criterion = CrossEntropyLabelSmooth(
+            1000, # num classes
```
Comment (katport) [inline examples/experimental/imagenet_nas_arch_pytorch/model_def.py:45]: hyperparameter the number of classes

File: examples/experimental/imagenet_nas_arch_pytorch/model_def.py [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+            bucket_name=bucket_name,
+            transform=train_transforms
+        )
+
+        train_queue = DataLoader(
+            train_data,
+            batch_size=self.context.get_per_slot_batch_size(),
+            shuffle=True,
+            pin_memory=True,
+            num_workers=16,
```
Comment (katport) [inline examples/experimental/imagenet_nas_arch_pytorch/model_def.py:82]: hyperparameter the num_workers in possibly data_config

File: examples/experimental/imagenet_nas_arch_pytorch/model_def.py [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+                ('max_pool_3x3', 2)
+            ], reduce_concat=range(2,6)
+        )
+        activation_function = activation_map[self.context.get_hparam("activation")]
+
+        model = NetworkImageNet(
+            genotype,
+            activation_function,
+            self.context.get_hparam("init_channels"),
+            1000, # number of classes
```
Comment (katport) [inline examples/experimental/imagenet_nas_arch_pytorch/model_def.py:144]: same as above

File: examples/experimental/imagenet_nas_arch_pytorch/model_def.py [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+            model.parameters(),
+            lr=self.context.get_hparam("learning_rate"),
+            momentum=self.context.get_hparam("momentum"),
+            weight_decay=self.context.get_hparam("weight_decay")
+        )
+        return self._optimizer
+
+    def create_lr_scheduler(self, optimizer):
+        self.scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(
+            optimizer, 300
```
Comment (katport) [inline examples/experimental/imagenet_nas_arch_pytorch/model_def.py:166]: set 300 to hyperparameter

File: examples/experimental/imagenet_nas_arch_pytorch/model_def.py [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+
+        if batch_idx==0 or self.last_epoch_idx < epoch_idx:
+            self.scheduler.step()
+            current_lr = self.scheduler.get_last_lr()[0]
+
+            if epoch_idx < 5:
+                lr = self.context.get_hparam("learning_rate")
+                for param_group in self._optimizer.param_groups:
+                    param_group["lr"] = lr * (epoch_idx + 1) / 5.0
+                print("Warming-up Epoch: {}, LR: {}".format(epoch_idx, lr * (epoch_idx + 1) / 5.0))
```
Comment (katport) [inline examples/experimental/imagenet_nas_arch_pytorch/model_def.py:183]: We may want to consider putting this into a LRschduler class so we can just pass in a LRschduler and have determined handle the scheduler for us. 

Did this produce the correct numbers in dtrain setting? I think we create a copy of the optimizer in horovod which means this would not be the same optimizer used for training. 
`            self.optimizer = hvd.DistributedOptimizer(
                self.optimizer,
                ....)`

The above LR scheduler solution should also solve this problem for us.

File: examples/experimental/imagenet_nas_arch_pytorch/model_def.py [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+        model.drop_path_prob = 0
+        start_time = time.time()
+
+        logits, logits_aux = model(input)
+        loss = self.criterion(logits, target)
+        if self.context.get_hparam("auxiliary"):
+            loss_aux = self.criterion(logits_aux, target)
+            loss += self.context.get_hparam("auxiliary_weight") * loss_aux
+        top1, top5 = accuracy(logits, target, topk=(1, 5))
+        #print("Time for batch: {}".format(time.time()-start_time))
```
Comment (katport) [inline examples/experimental/imagenet_nas_arch_pytorch/model_def.py:197]: remove comment

File: examples/experimental/imagenet_nas_arch_pytorch/model_def.py [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+                lr = self.context.get_hparam("learning_rate")
+                for param_group in self._optimizer.param_groups:
+                    param_group["lr"] = lr * (epoch_idx + 1) / 5.0
+                print("Warming-up Epoch: {}, LR: {}".format(epoch_idx, lr * (epoch_idx + 1) / 5.0))
+            else:
+                print("Epoch: {} lr {}".format(epoch_idx, current_lr))
+
+        input, target = batch
+        model.drop_path_prob = 0
+        start_time = time.time()
```
Comment (katport) [inline examples/experimental/imagenet_nas_arch_pytorch/model_def.py:189]: remove time

File: examples/experimental/imagenet_nas_arch_pytorch/data.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+    def __init__(self, split, bucket_name, transform=None):
+        """
+        Args:
+            split: train or validation split to return the right dataset
+            directory: root directory for imagenet where "train" and "validation" folders reside
+        """
+        assert split in ["train", "validation"], "split {} not in (train, validation)".format(split)
+        self._split = split
+        self._bucket_name = bucket_name
+        self._target_dir = '/mnt/data'
```
Comment (katport) [inline examples/experimental/imagenet_nas_arch_pytorch/data.py:57]: add to data_config

File: examples/experimental/imagenet_nas_arch_pytorch/data.py [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+def create_file_dirs(target_path):
+    destination_dir = target_path[0:target_path.rfind('/')]
+    if not os.path.exists(destination_dir):
+        try:
+            os.makedirs(destination_dir)
+        except:
+            assert os.path.exists(destination_dir)
+            pass
+
+def download_entire_imagenet(data_dir='/mnt/data'):
```
Comment (katport) [inline examples/experimental/imagenet_nas_arch_pytorch/data.py:39]: Is this used still? Can't seem to find where it's called.

File: examples/experimental/imagenet_nas_arch_pytorch/data.py [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+            self._storage_client = storage.Client()
+            self._bucket = self._storage_client.bucket(bucket_name)
+            # When the dataset is first initialized, we'll loop through to catalogue the classes (subdirectories)
+            # This step might take a long time.
+            self._imgs_paths = list_blobs(self._storage_client, self._bucket_name, prefix=self._source_dir)
+
+            self._labels = []
+            self._subdir_to_class = {}
+            class_count = 0
+            for path in self._imgs_paths:
```
Comment (katport) [inline examples/experimental/imagenet_nas_arch_pytorch/data.py:70]: when we loop through to get the path "names" can we also loop through to get some of these counts to prevent looping through 1mil + images?

File: examples/experimental/imagenet_nas_arch_pytorch/data.py [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+                    self._subdir_to_class[sub_dir] = class_count
+                    class_count += 1
+                self._labels.append(self._subdir_to_class[sub_dir])
+            print("There are {} records in dataset.".format(len(self._imgs_paths)))
+            self._streaming = True
+
+    def __len__(self):
+        # Return some length if using randomly generated data.
+        if self._bucket_name is None:
+            return 1024*100
```
Comment (katport) [inline examples/experimental/imagenet_nas_arch_pytorch/data.py:82]: set as data_config

File: examples/experimental/imagenet_nas_arch_pytorch/data.py [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+        return len(self._imgs_paths)
+
+
+    def __getitem__(self, idx):
+        if self._bucket_name is None:
```
Comment (katport) [inline examples/experimental/imagenet_nas_arch_pytorch/data.py:87]: add some comments

File: examples/experimental/imagenet_nas_arch_pytorch/data.py [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+            if not os.path.exists(target_path):
+                create_file_dirs(target_path)
+                print('downloading...')
+                img_str = download_gcs_blob_with_backoff(blob)
+                with open(target_path,'wb') as f:
+                    f.write(img_str)
+            else:
+                with open(target_path, 'rb') as f:
+                    img_str = f.read()
+
```
Comment (katport) [inline examples/experimental/imagenet_nas_arch_pytorch/data.py:105]: lets add that there is an option to pick stream and download or just download. This can get a bit confusing here.

File: examples/experimental/imagenet_nas_arch_pytorch/data.py [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+        if self._transform is not None:
+            img = self._transform(img)
+
+        return img, self._labels[idx]
+
+if __name__=="__main__":
+    dataset = ImageNetDataset('validation', None)
+    print(len(dataset))
+    print(dataset[10])
```
Comment (katport) [inline examples/experimental/imagenet_nas_arch_pytorch/data.py:118]: is this for testing? If so let's remove it.

File: examples/experimental/imagenet_nas_arch_pytorch/README.md [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+* **utils.py**: Contains special functions used during training.
+* **data.py**: Contains the data loading code for the model.
+* **distributed.yaml**: Experiment configuration for distributed training with fixed hyperparameter setting.
+
+### Data
+The data for ImageNet is the ILSVRC2012 version of the dataset, which is available [here](http://www.image-net.org/).  The code assumes the data is available in a Google Cloud Storage bucket; in the event a bucket is not provided, randomly generated data is used.
+
+### To Run
+   *Prerequisites*:  
+      A Determined Google Cloud Compute cluster must be installed in order to run this example.  Please follow the directions [here](https://docs.determined.ai/latest/how-to/install-main.html) in order to install.
```
[the remark this answers is not in the document]
Comment (liamcli) [reply examples/experimental/imagenet_nas_arch_pytorch/README.md:19]: I changed to: Contains the model training code expected by Determined.

File: examples/experimental/imagenet_nas_arch_pytorch/README.md [hunk 19]
```diff
@@ -0,0 +1,28 @@
+# Training state-of-the-art architecture on ImageNet
+In this example, we train a state-of-the-art architecture discovered by a recently published neural architecture search algorithm called GAEA on the ImageNet dataset.  Please see the original paper by [Li et al.](https://arxiv.org/abs/2004.07802) for more information about the underlying NAS algorithm used to discover this architecture.
+
+The training routine is largely based on the one used by [PC-DARTS](https://github.com/yuhuixu1993/PC-DARTS/blob/master/train_imagenet.py) with additional tricks for improved performance.  In particular, we use the swish activation and additional squeeze and excite modules in place of the auxiliary tower used in the PC-DARTS training routine.  
+
+### Folders and Files
+* **model_def.py**: Ports model to the Determined PyTorchTrial interface.
```
[the remark this answers is not in the document]
Comment (liamcli) [reply examples/experimental/imagenet_nas_arch_pytorch/README.md:7]: I changed to: Contains the model training code expected by Determined.

File: examples/experimental/imagenet_nas_arch_pytorch/README.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (liamcli) [reply examples/experimental/imagenet_nas_arch_pytorch/README.md:19]: We decided at AMLE sync today that we can push this out without adding AWS.  The main consideration is that we don't currently support spot instances on AWS and this experiment can get expensive.  We can always come back to add it if someone requests.

File: examples/experimental/imagenet_nas_arch_pytorch/data.py (hunk 13 above)
Comment (liamcli) [reply examples/experimental/imagenet_nas_arch_pytorch/data.py:70]: good suggestion, i combined into one loop

File: examples/experimental/imagenet_nas_arch_pytorch/data.py (hunk 14 above)
Comment (liamcli) [reply examples/experimental/imagenet_nas_arch_pytorch/data.py:82]: i don't think it's necessary to include the size of randomly generated data as a config parameter.  i could just as well put np.random.int

File: examples/experimental/imagenet_nas_arch_pytorch/model_def.py (hunk 8 above)
Comment (liamcli) [reply examples/experimental/imagenet_nas_arch_pytorch/model_def.py:183]: This does not create a problem for dtrain because every instance of optimizer calls this (as evidenced by the log printing the warmup lr 8 times).  It might be cleaner to implement in a custom scheduler but the original PC-DARTS codebase also coded the warmup lr like this so I think it's okay to keep as is.

**Comment by aaron276h** [comment]
@liamcli We have recently made a push to add all our example models to be a part of our nightly/weekly tests, running for one or two steps rather than to convergence.

**Comment by liamcli** [comment]
> @liamcli We have recently made a push to add all our example models to be a part of our nightly/weekly tests, running for one or two steps rather than to convergence.

Added capability test to `test_experimental.py` and `nightly/test_capability.py` using randomly generated data.
